### PR TITLE
MON-6280: Add a special variable for trap OID

### LINKED ISSUE #9044: Sending SNMP trap 'Enterprise OID' to a 'preexec' command as a variable
Hi!

I've been testing Centreon and I'm specially interested in passive monitoring (we manage a lot of SNMP traps).
I've been learning about how Centreon manages SNMP traps and I know trap variables can be passed to a preexec command using 'macros' like @{NUMERIC_OID}. 

But is there any 'macro' for trap Enterprise OID? It would be very useful for us to be able to pass the trap Enterprise OID to a preexec script in order to unify traps managemenent. According to https://docs.centreon.com/current/en/monitoring/passive-monitoring/create-snmp-traps-definitions.html, I pass to my preexec script a list of variables, for instance @TRAPOUTPUT@ or @{NUMERIC_OID}, but what about the Enterprise OID? What about the OID I use to define the trap in Centreon? Can I get it? Is there any contextual variable that keeps it?

Thankd in advance!

Albert

### PR DESCRIPTION
**Pull request by garnier-quentin**
## Description

Fix  #9044

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [X] 20.04.x
- [X] 20.10.x
- [X] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

The special variable ```$o``` can be used. You can try to add it in trap output. The trap OID should be displayed. 

## Checklist

- [X] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by ahurtadob** [comment]
Hello,
I'm using a 20.04 version. I replaced centreontrapd.pm. It does not work. No $o is showed in output neither in arguments passed to a command.
I'm going to start again and repeat all steps. I'm new in Centreon and maybe I'm doing something wrong
Thanks!

**Comment by garnier-quentin** [comment]
You only need to replace lines with $o

**Comment by ahurtadob** [comment]
Hi again,
A restart needed. Now it works perfectly. I can see the OID in output and in arguments passed to a command.
Thanks a lot!
Albert

**Comment by garnier-quentin** [comment]
Thanks for the feedback!

**Review by cgagnaire — DISMISSED** [review]

**Comment by cgagnaire** [comment]
@kduret Can you look at this one please, mostly lines 892-893?
Thanks!

File: lib/perl/centreon/script/centreontrapd.pm [hunk 1]
```diff
@@ -853,38 +857,41 @@ sub substitute_host_macro {
...
+    $str =~ s/\$r/$self->{trap_data}->{var}->[0]/g;
+
     # $aR, $ar (IP Adress)
-    $str =~ s/\$aR/${$self->{trap_data}->{var}}[1]/g;
-    $str =~ s/\$ar/${$self->{trap_data}->{var}}[1]/g;
-    
+    $str =~ s/\$aR/$self->{trap_data}->{var}->[1]/g;
+    $str =~ s/\$ar/$self->{trap_data}->{var}->[1]/g;
+
+    # $o (Trap OD)
```
Comment (kduret) [inline lib/perl/centreon/script/centreontrapd.pm:892]:
```suggestion
    # $o (Trap OID)
```

**Review by kduret — APPROVED** [review]
LGTM
Probably needs documentation here : https://docs.centreon.com/current/en/monitoring/passive-monitoring/monitoring-with-snmp-traps.html

**Review by kduret — APPROVED** [review]

**Review by cgagnaire — APPROVED** [review]

**Pull request merged**